### PR TITLE
Auction

### PR DESCRIPTION
**Pull request by viken33**
## The Problem

Auction MVP
[Specs](https://www.notion.so/Auction-MVP-87ba3006ca4648b29c24c6440b4855ea#5b70cdadf7a34c69bb13dc01c103f47b)

## The Solution

This PR is a check-in send to see if we are on the right track and if changes on different modules/files are according to MVP design. Its on auction branch on https://github.com/1Hive/vector.git

Update:
Added an Auction Stack to vector in order to test multi router / auction flow
Basic functionality working, can be tested from server-node RPC but in order to run an auction:
      - All channels must be setup first. This is basically because actual method in [fee service from router](https://github.com/connext/vector/blob/97ffc15ced65da0a33107b98de53e825f589b745/modules/router/src/services/fees.ts#L31) takes the channels as params. A different method is needed to calculate fees based only on PUBIDs (lookup channels, calculate deployment fees, etc). Hence, if channels are not present router throws an error and doesn't respond to auction message.
      - Auction must be in the same chain, otherwise routers throw an error about not being able to calculate swapRates/AMM rates. Not sure why.

Main additions:
        Start-auction script to deploy 3 routers
        3 different router.config files to have different fee rates
        Makerfile changes ( `make start-auction` )

### CODE REVIEW

File: modules/engine/src/index.ts [hunk 1]
```diff
@@ -67,6 +69,8 @@ export class VectorEngine implements IVectorEngine {
 
   private readonly restoreLocks: { [channelAddress: string]: string } = {};
 
+  private auctionResponses: Array<NodeResponses.RunAuction> = [];
```
Comment (viken33) [inline modules/engine/src/index.ts:72]: global store for auction responses, maybe it should be a variable inside scope of runAuction method only

File: modules/engine/src/index.ts [hunk 2]
```diff
@@ -1525,6 +1529,96 @@ export class VectorEngine implements IVectorEngine {
...
+      senderPublicIdentifier: this.publicIdentifier,
+      senderAssetId: params.assetId,
+      senderChainId: params.chainId,
+      receiverPublicIdentifier: params.recipient,
+      receiverAssetId: params.recipientAssetId,
+      receiverChainId: params.recipientChainId,
+    };
+    this.evts[EngineEvents.RUN_AUCTION_EVENT].post(payload);
+
+    const inbox = getRandomBytes32();
```
Comment (viken33) [inline modules/engine/src/index.ts:1557]: to be changed. Couldn't find a service to generate a random inbox with the format "_INBOX.xxxxxxx"

File: modules/engine/src/index.ts [hunk 3]
```diff
@@ -1525,6 +1529,96 @@ export class VectorEngine implements IVectorEngine {
...
+      this.logger.info(this.auctionResponses[lowestFeeIndex], "Chosen Router");
+      return Result.ok(this.auctionResponses[lowestFeeIndex]);
+    } catch (err) {
+      return Result.fail(
+        new RpcError(RpcError.reasons.InvalidParams, "", this.publicIdentifier, {
+          invalidParamsError: validate.errors?.map((e) => e.message).join(","),
+          invalidParams: params,
+        }),
+      );
+    } finally {
```
Comment (viken33) [inline modules/engine/src/index.ts:1617]: TODO: Unsubscribe from inbox

File: modules/router/src/listener.ts [hunk 4]
```diff
@@ -822,6 +822,226 @@ export async function setupListeners(
       );
     }
   });
+  // TODO Refactoring, repeated code
+  await messagingService.onReceiveStartAuction(routerSigner.publicIdentifier, async (request, from, inbox) => {
```
Comment (viken33) [inline modules/router/src/listener.ts:826]: To be changed. I used code from other listeners. Would depend on new methods to calculate fees/swapRates

File: modules/types/src/engine.ts [hunk 5]
```diff
@@ -117,6 +117,18 @@ export type WithdrawalReconciledPayload = {
 export const RESTORE_STATE_EVENT = "RESTORE_STATE_EVENT";
 export type RestoreStatePayload = SetupPayload;
 
+// Emitted on Auction
+export const RUN_AUCTION_EVENT = "RUN_AUCTION_EVENT";
+export type RunAuctionPayload = {
```
Comment (viken33) [inline modules/types/src/engine.ts:122]: to be reviewed about optional/mandatory params

File: modules/types/src/schemas/engine.ts [hunk 6]
```diff
@@ -167,6 +167,17 @@ const WithdrawParamsSchema = Type.Object({
   initiatorSubmits: Type.Optional(Type.Boolean()),
 });
 
+// Run Auction Params
+
+const RunAuctionParamsSchema = Type.Object({
```
Comment (viken33) [inline modules/types/src/schemas/engine.ts:172]: to be reviewed: should params be mandatory or optional ?

File: modules/utils/src/messaging.ts [hunk 7]
```diff
@@ -170,6 +171,13 @@ export class NatsBasicMessagingService implements IBasicMessaging {
     await this.connection!.publish(subject, toPublish);
   }
 
+  public async publishUniqueInbox(subject: string, data: any, inbox: string): Promise<void> {
```
Comment (viken33) [inline modules/utils/src/messaging.ts:174]: helper method to broadcast message with replayTo param

**Review by viken33 — COMMENTED** [review]
added some comments

File: modules/router/src/listener.ts [hunk 8]
```diff
@@ -822,6 +822,226 @@ export async function setupListeners(
...
+    let recipientChannel = recipientChannelRes.getValue() as FullChannelState | undefined;
+    if (!recipientChannel) {
+      const placeholder = await getEmptyChannel(recipient, chainId);
+      if (placeholder.isError) {
+        await messagingService.respondToAuctionMessage(inbox, Result.fail(placeholder.getError()!));
+        return;
+      }
+      recipientChannel = placeholder.getValue();
+    }
+    const feeRes = await calculateFeeAmount(
```
Comment (rhlsthrm) [inline modules/router/src/listener.ts:974]: May need to change this to consider the case where channel does not exist.

File: modules/router/src/listener.ts [hunk 9]
```diff
@@ -822,6 +822,226 @@ export async function setupListeners(
...
+      return;
+    }
+
+    const getEmptyChannel = async (
+      counterparty: string,
+      chainId: number,
+    ): Promise<Result<FullChannelState, QuoteError>> => {
+      const alice = getSignerAddressFromPublicIdentifier(routerSigner.publicIdentifier);
+      const bob = getSignerAddressFromPublicIdentifier(counterparty);
+      const channelAddress = await chainReader.getChannelAddress(
```
Comment (rhlsthrm) [inline modules/router/src/listener.ts:913]: Nice work on this!

File: modules/router/src/listener.ts [hunk 10]
```diff
@@ -822,6 +822,226 @@ export async function setupListeners(
...
+      recipient,
+      recipientChainId,
+      recipientAssetId,
+      fee: fee.toString(),
+      expiry: (Date.now() + (getConfig().feeQuoteExpiry ?? DEFAULT_FEE_EXPIRY)).toString(), // valid for next 2 blocks
+    };
+    const toSign = hashTransferQuote(quote);
+    try {
+      const signature = await routerSigner.signMessage(toSign);
+      await messagingService.respondToAuctionMessage(
```
Comment (rhlsthrm) [inline modules/router/src/listener.ts:1018]: Do we need to send the full quote so that we have the signature?

File: modules/server-node/examples/0-config.http [hunk 11]
```diff
@@ -71,3 +73,11 @@ Content-Type: application/json
 {
   "index": 0
 }
+
+### CREATE DAVE
+POST {{daveUrl}}/node
```
Comment (rhlsthrm) [inline modules/server-node/examples/0-config.http:78]: Revert this change please.

File: modules/types/src/schemas/engine.ts (hunk 6 above)
Comment (rhlsthrm) [reply modules/types/src/schemas/engine.ts:172]: Let's make all of them mandatory just to make things easier.

**Comment by rhlsthrm** [comment]
This looks AMAZING @viken33 ! A couple missing things:
- Add auction method to https://github.com/connext/vector/blob/main/modules/types/src/node.ts#L14.
- The above will create build errors where the appropriate methods need to be added in the REST client and the browser node.

**Review by rhlsthrm — CHANGES_REQUESTED** [review]

File: modules/engine/src/index.ts (hunk 2 above)
Comment (viken33) [reply modules/engine/src/index.ts]: changed

File: modules/engine/src/index.ts (hunk 3 above)
Comment (viken33) [reply modules/engine/src/index.ts:1671]: added

File: modules/types/src/engine.ts (hunk 5 above)
Comment (viken33) [reply modules/types/src/engine.ts:122]: changed to mandatory

File: modules/types/src/messaging.ts [hunk 12]
```diff
@@ -184,4 +188,17 @@ export interface IMessagingService extends IBasicMessaging {
...
   ): Promise<void>;
+
+  publishStartAuction(
+    to: string,
+    from: string,
+    data: Result<EngineParams.RunAuction, NodeError>,
+    inbox: string,
+  ): Promise<void>;
+
+  onReceiveAuctionMessage(
```
[the remark this answers is not in the document]
Comment (viken33) [reply modules/types/src/messaging.ts:199]: unsubscribed in finally block on runAuction

File: modules/types/src/schemas/engine.ts (hunk 6 above)
Comment (viken33) [reply modules/types/src/schemas/engine.ts:172]: changed to mandatory

File: Makefile [hunk 13]
```diff
@@ -49,6 +49,7 @@ router-prod: node-prod router-img
 
 duet: messaging server-node-js
 trio: messaging server-node-js router-js
+auction: messaging server-node-js router-js
```
[the remark this answers is not in the document]
Comment (viken33) [reply Makefile:52]: I left this, found it helpful but discard it if you prefer

File: modules/engine/src/index.ts [hunk 14]
```diff
@@ -1525,6 +1529,96 @@ export class VectorEngine implements IVectorEngine {
...
+    const inbox = getRandomBytes32();
+    const from = this.signer.publicIdentifier;
+
+    // Call onReceiveAuctionMessage to listen on unique INBOX and collect responses for 5 seconds (will tweak and tune this number).
+    // Maybe something like wait for 5 responses or 5 seconds? Watch out for race conditions of setting listener after message is already sent.
+    try {
+      //Call publishStartAuction with provided data.
+      this.messaging.publishStartAuction(from, from, Result.ok(params), inbox);
+
+      function waitForRespones(t) {
```
[the remark this answers is not in the document]
Comment (viken33) [reply modules/engine/src/index.ts]: changed

File: modules/engine/src/index.ts [hunk 15]
```diff
@@ -1525,6 +1529,96 @@ export class VectorEngine implements IVectorEngine {
...
+      });
+
+      // wait for 5 responses or 3 secs
+      if (this.auctionResponses.length < 5) {
+        await waitForRespones(3000);
+      }
+
+      this.logger.info(this.auctionResponses, "Router Responses");
+
+      if (this.auctionResponses.length == 0) {
```
[the remark this answers is not in the document]
Comment (viken33) [reply modules/engine/src/index.ts]: ok changed

File: modules/router/src/listener.ts (hunk 10 above)
Comment (viken33) [reply modules/router/src/listener.ts:1018]: Signing only the quote

File: modules/server-node/examples/0-config.http (hunk 11 above)
Comment (viken33) [reply modules/server-node/examples/0-config.http]: yeah sorry, it was for testing. 
I reverted the examples to originals, and added an 'auction.http' for testing. It can be discarded

**Review by rhlsthrm — CHANGES_REQUESTED** [review]
This looks great, thanks for addressing all the changes!

A couple of final things we need:

- [ ] Unit tests
- [ ] Feature flag. We need to add a config item so that we can merge this into our prod branch and disable the feature until we fully test it.

File: modules/engine/src/testing/index.spec.ts [hunk 16]
```diff
@@ -803,6 +803,181 @@ describe("VectorEngine", () => {
           overrides: { method: "chan_defundTransfer", params: { transferId: invalidAddress } },
           error: malformedTransactionId,
         },
+        {
+          name: "chan_runAuction malformed parameter amount",
+          overrides: {
```
Comment (viken33) [inline modules/engine/src/testing/index.spec.ts:808]: added rpc param unit tests

File: modules/router/src/test/services/messaging.spec.ts [hunk 17]
```diff
@@ -57,4 +58,44 @@ describe("messaging.ts", () => {
     expect(res.isError).to.be.false;
     expect(res.getValue()).to.be.deep.eq(configResponse);
   });
+
+  // TODO: replace hardcoded swapRate
+  it("should properly respond with auction response when requested", async () => {
+    const auctionResponse = {
+      routerPublicIdentifier: router.publicIdentifier,
```
Comment (viken33) [inline modules/router/src/test/services/messaging.spec.ts:65]: added messaging test for auction related messages

**Comment by viken33** [comment]
Added unit tests where I found them to be relevant for auction (rpc method on engine, listener on router).
If there are more tests we should add, please point me to the tests specs files or folders and I'll them :+1:

**Review by rhlsthrm — APPROVED** [review]
